eth/watchers/senderwatcher_test.go

```diff
@@ -0,0 +1,595 @@
+	assert.Zero(info.Reserve.Cmp(big.NewInt(5)))
+	info = sw.senders[sender]
+	assert.Zero(info.Deposit.Cmp(big.NewInt(10)))
+	assert.Zero(info.Reserve.Cmp(big.NewInt(5)))
+}
+
+func TestSenderWatcher_Clear(t *testing.T) {
+	assert := assert.New(t)
+	stubClientSenderInfo := &pm.SenderInfo{
+		Deposit: big.NewInt(10),
+		Reserve: big.NewInt(5),
+	}
+	sw := &SenderWatcher{
+		senders: make(map[ethcommon.Address]*pm.SenderInfo),
+		lpEth: &eth.StubClient{
+			SenderInfo: stubClientSenderInfo,
+		},
+	}
+	info := &pm.SenderInfo{
+		Deposit:      big.NewInt(1000000000000000000),
+		Reserve:      big.NewInt(1000000000000000000),
+		ReserveState: pm.NotFrozen,
+	}
+	sender := pm.RandAddress()
+
+	sw.setSenderInfo(sender, info)
+
+	getInfo, err := sw.GetSenderInfo(sender)
+	assert.Nil(err)
+	assert.Equal(info, getInfo)
+
+	sw.Clear(sender)
+
+	getInfo = sw.senders[sender]
+	assert.Nil(getInfo)
+	// on calling GetSenderInfo should be back to RPC stub values
+	getInfo, err = sw.GetSenderInfo(sender)
+	assert.Nil(err)
+	assert.Equal(getInfo, stubClientSenderInfo)
+}
+
+func TestSenderWatcher_WatchAndStop(t *testing.T) {
+	assert := assert.New(t)
+	lpEth := &eth.StubClient{}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	go sw.Watch()
+	time.Sleep(2 * time.Millisecond)
+	sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+	assert.True(watcher.sub.unsubscribed)
+}
+
+func TestSenderWatcher_HandleLog(t *testing.T) {
+	lpEth := &eth.StubClient{}
+	watcher := &stubBlockWatcher{}
+	rw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	require.Nil(t, err)
+
+	assert := assert.New(t)
+
+	// Test unknown event
+	log := newStubBaseLog()
+	log.Topics = []ethcommon.Hash{ethcommon.BytesToHash([]byte("foo"))}
+
+	err = rw.handleLog(log)
+	assert.Nil(err)
+}
+
+func TestFundDepositEvent(t *testing.T) {
+	assert := assert.New(t)
+	require := require.New(t)
+	startDeposit := big.NewInt(10)
+	lpEth := &eth.StubClient{
+		SenderInfo: &pm.SenderInfo{
+			Deposit: big.NewInt(10),
+			Reserve: big.NewInt(5),
+		},
+	}
+	watcher := &stubBlockWatcher{}
+	sw, err := NewSenderWatcher(stubTicketBrokerAddr, watcher, lpEth)
+	assert.Nil(err)
+
+	header := defaultMiniHeader()
+	newDepositEvent := newStubDepositFundedLog()
+	header.Logs = append(header.Logs, newDepositEvent)
+
+	blockEvent := &blockwatch.Event{
+		Type:        blockwatch.Added,
+		BlockHeader: header,
+	}
+
+	go sw.Watch()
+	defer sw.Stop()
+	time.Sleep(2 * time.Millisecond)
+
+	// If map entry is nil don't handle event
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	_, ok := sw.senders[stubSender]
+	assert.False(ok)
+
+	// if block removed, make RPC call
+	blockEvent.Type = blockwatch.Removed
+	watcher.sink <- []*blockwatch.Event{blockEvent}
+	time.Sleep(2 * time.Millisecond)
+	info, err := sw.GetSenderInfo(stubSender)
+	assert.Nil(err)
+	assert.Zero(info.Deposit.Cmp(startDeposit))
```

Comment:
This test would still pass even if we remove L153 since `GetSenderInfo()` will cache the sender info for `stubSender` if the current internal map entry for `stubSender` is nil. Perhaps we could access the internal map directly to make sure we're not populating the cache via a call to `GetSenderInfo()`? Same comment applies to the rest of the tests in this file where we do this.

Reply:
fixed in 4530eff